designs/0031-deprecation-schedule.md

```diff
@@ -0,0 +1,254 @@
+   a major version bump dramatically increases the friction for updating. While
+   some may find it preferable to have to update their models only once, the
+   small requirement of (at most) 1-2 changes per version is likely preferable
+   to the many required changes a major version would bring.
+
+2. Do not allow the deprecation of features. If we do not plan on actually
+   removing deprecations, emitting warnings to the user is simply annoying and
+   nonproductive.
+
+   We instead decide to strongly commit to backwards compatibility in all
+   cases - no new language keywords, the preservation of current syntax ad
+   infinitum, etc. This has benefits to the end user - a Stan program written
+   today would still work (assuming we were actually able to meet this high
+   standard) on a version of Stan released years from now. The downside is that
+   the codebase would bloat over time, and we will run into more and more
+   instances where the correct behavior or error message is indiscernible in
+   practice. Tools which build around Stan in particular have trouble with our
+   treatment of some things as valid identifiers and valid keywords at the same
+   time - syntax highlighting especially suffers:
+   ```stan
+       real<lower=0> x; // notice lower is highlighted
+       real<offset=0> y; // notice offset is not
+       int offset; // this is (partially) the reason why
+   ```
+
+# Prior art
+[prior-art]: #prior-art
+
+- The original inspiration for this schedule comes from the Python library
+  NumPy, with their deprecation policy laid out in
+  [NEP-23](https://numpy.org/neps/nep-0023-backwards-compatibility.html#implementing-deprecations-and-removals).
+  The NumPy project provides a similar example of a large open-source code base
+  whose minor version number has now exceeded 20 (meaning a large number of
+  changes since the last 'breaking'/major version).
+- Python itself defines a similar policy in
+  [PEP-387](https://www.python.org/dev/peps/pep-0387/) which requires warnings
+  emitted for 2 minor versions or 1 major version (e.g, a feature deprecated in
+  3.10 could be removed in 3.12 **or** 4.0)
+- The Java language clarified its [deprecation
+  policy](https://docs.oracle.com/javase/9/core/enhanced-deprecation1.htm#JSCOR-GUID-23B13A9E-2727-42DC-B03A-E374B3C4CE96)
+  in Java 9 (which is technically a minor version, 1.9, for the JDK). The main
+  motivation for clarification was the response of users being unsure if a
+  deprecated feature would ever actually be removed. There was a
+  [study](https://doi.org/10.1007/S10664-019-09713-W) of this change published
+  in 2019. It contains the following relevant passage and some interesting
+  discussion of user reactions to a deprecation warning.
+  > By not removing deprecated features from an API after a transition period
+  > has passed, API producers and the Java JDK developers themselves have
+  > cheapened the meaning of deprecation.
```

Comment:
[comment]
Uh oh.  We've also cheapened the term!

Reply:
It's a really interesting article, in part because it seems like nearly _everybody_ cheapens the term